docs/proposal/20200506-single-controller-multitenancy.md

```diff
@@ -0,0 +1,869 @@
+
+<a name="NFR8">NFR10.</a> Unit tests MUST exist for all credential provider code.
+
+<a name="NFR8">NFR11.</a> e2e tests SHOULD exist for all credential provider code.
+
+<a name="NFR8">NFR12.</a> Credential provider code COULD be audited by security engineers.
+
+### Implementation Details/Notes/Constraints
+
+The current implementation of CAPA requests a new AWS EC2 and ELB service per cluster and per machine and stores these
+in fields on the ClusterScope struct. ClusterScopes are reference values which are created per-reconciliation:
+
+``` go
+type ClusterScope struct {
+  logr.Logger
+  client      client.Client
+  patchHelper *patch.Helper
+
+  AWSClients
+  Cluster    *clusterv1.Cluster
+  AWSCluster *infrav1.AWSCluster
+}
+```
+
+The field AWSClients holds the actual AWS service clients, and is defined like so:
+
+``` go
+type AWSClients struct {
+  EC2             ec2iface.EC2API
+  ELB             elbiface.ELBAPI
+  SecretsManager  secretsmanageriface.SecretsManagerAPI
+  ResourceTagging resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI
+}
+```
+
+The signatures for the functions which create these instances are as follows:
+
+``` go
+func NewClusterScope(params ClusterScopeParams) (*ClusterScope, error) {
+  ...
+  return &ClusterScope{
+    ...
+  }, nil
+}
+```
+
+#### Proposed Changes
+
+This proposal borrows heavily from the Service APIs specification.
+
+In the initial implementation, all new resources will be scoped at the cluster level, this is to enable delegation of
+AWS accounts whilst preventing privilege escalation as per [FR4](#FR4). Reasons for this are documented in the
+[alternatives](#alternatives) section.
+
+##### Cluster API Provider AWS v1alpha3 types
+
+<strong><em>Changed Resources</strong></em>
+* `AWSCluster`
+
+<strong><em>New Resources</strong></em>
+
+<em>Cluster scoped resources</em>
+
+* `AWSClusterStaticPrincipal` represents a static AWS tuple of credentials.
+* `AWSClusterRolePrincipal` represents an intent to assume an AWS role for cluster management.
+
+<em>Namespace scoped resources</em>
+
+* `AWSServiceAccountPrincipal` represents the use of a Kubernetes service account for access
+  to AWS using federated identity.
+
+<strong><em>Changes to AWSCluster</em></strong>
+
+A new field is added to the `AWSClusterSpec` to reference an principal. We intend to use `corev1.LocalObjectReference` in
+order to ensure that the only objects that can be references are either in the same namespace or are scoped to the
+entire cluster.
+
+``` go
+type  AWSClusterSpec  struct {
+  PrincipalRef corev1.LocalObjectReference `json:"principalRef,omitempty"`
+  AccountID string
+```
+
+An example usage would be:
+
+``` yaml
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AWSCluster
+metadata:
+  name: "test"
+  namespace: "test"
+spec:
+  region: "eu-west-1"
+  principalRef:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+    kind: AWSClusterRolePrincipal
+    name: test-account-role
```

Comment:
In this example the reference has also `apiVersion` and `kind`, although `LocalObjectReference` doesn't have those field, and `TypedLocalObjectReference` shouldn't be used for non-core types or it has been deprecated (cc @detiber @ncdc).

Maybe this is a good time to introduce our own reference in Cluster API and use it? Or keep using ObjectReference for now but force the namespace to be the same in code?

Reply:
The conclusion of this was to await a final determination from @fabriziopandini as to whether `clusterctl move` ever relies on these references in infrastructure providers and needs them typed a particular way, or are OwnerReferences sufficient.

If setting OwnerReferences are sufficient, then we will have a local PrincipalReference type instead of using corev1.*, and additionally move Principal types into their own API group.

Reply:
Had a discussion here https://kubernetes.slack.com/archives/C8TSNPY4T/p1590063805192000 , after which I have updated the proposal to ensure we add the principal resource as an owner reference to the AWSCluster. A new section entitled `clusterctl changes` documents that we need to add support for the move of cluster-scoped resources.

Based on that, I'm in favour of having a specific locally typed reference field.